packages/@react-stately/data/src/useAsyncList.ts

```diff
@@ -243,34 +272,50 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
         selectedKeys: data.selectedKeys,
         sortDescriptor: action.sortDescriptor ?? data.sortDescriptor,
         signal: abortController.signal,
-        cursor: action.type === 'loadingMore' ? data.cursor : null
+        cursor: action.type === 'loadingMore' ? data.cursor : null,
+        filterText: data.filterText
       });
       dispatch({type: 'success', ...response, abortController});
     } catch (e) {
       dispatch({type: 'error', error: e, abortController});
     }
   };
 
+  // Handle initial load
   useEffect(() => {
     dispatchFetch({type: 'loading'}, load);
   // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  // Reload on changes to filter text but only if performing server side filtering
+  // Kinda weird since it will trigger a reload when user selects a option
+  useEffect(() => {
+    if (!filterFn) {
+      dispatchFetch({type: 'loading'}, load);
+    }
+  }, [data.filterText, filterFn, load, dispatchFetch]);
```

Comment:
Tried to do this outside of `useAsyncList` via calling `list.reload` in `onInputChange` but this means `filterText` is a step behind due to `setFilter` being called in the same `onInputChange`

Also kinda weird that a reload will get triggered when the user selects an option since that changes the filter text

Reply:
I think you could get rid of this if you overwrite `setFilterText` to perform a new action